File: src/generate/generate.ts
import { join } from "node:path";
import type { GlobalResolution } from "../core/interpolate.ts";
import { expandAll } from "../core/interpolate.ts";
import type { PlanIssue } from "../core/plan.ts";
import { writeFileAtomic } from "../io/write.ts";
import type { Registry } from "../registry/types.ts";
import type { VaultSession } from "../vault/provider.ts";
import { disclaimerHeader, hasOwnershipMarker } from "./ownership.ts";
import { envTargets } from "./paths.ts";
import type { RenderEntry } from "./render.ts";
import { renderEnvContent, renderExampleContent, splitSecrets } from "./render.ts";

export interface GenerateOpts {
  vault?: string;
  consumer?: string;
}

export interface GeneratePreview {
  writes: { path: string; content: string }[];
  unchanged: string[];
  refused: string[]; // existing files without the ownership marker
  warnings: PlanIssue[];
}

export function globalsFor(registry: Registry, vault: string): Map<string, GlobalResolution> {
  const out = new Map<string, GlobalResolution>();
  for (const [name, def] of Object.entries(registry.globals)) {
    const v = def.values[vault];
    if (v === undefined) continue;
    out.set(name, v.source === "static" ? { kind: "static", value: v.value } : { kind: "runtime" });
  }
  return out;
}

// Vaults a generate over these options will read from (env targets only;
// compose adds the selected vault, which envTargets already covers for
// single-mode consumers — the CLI unions in the compose vault when needed).
export function vaultsNeeded(registry: Registry, opts: GenerateOpts): string[] {
  return [...new Set(envTargets(registry.consumers, registry.defaults, opts).map((t) => t.vault))].sort();
}

// All wired entries for one (consumer, vault) scope, values fetched and
// interpolation-expanded. Missing values render empty + MISSING_VALUE warning.
export async function scopeEntries(
  registry: Registry,
  consumer: string,
  vault: string,
  session: VaultSession,
  warnings: PlanIssue[],
): Promise<RenderEntry[]> {
  const raw = new Map<string, string>();
  const meta: { name: string; disabled: boolean }[] = [];
  for (const [name, def] of Object.entries(registry.variables)) {
    const entry = def.vaultMapping[vault]?.[consumer];
    if (entry === undefined) continue;
    const value = await session.get(entry.key);
    if (value === undefined) {
      warnings.push({
        code: "MISSING_VALUE",
        message: `"${name}" has no value in vault "${vault}" (consumer "${consumer}") — rendered empty`,
      });
    }
    raw.set(name, value ?? "");
    meta.push({ name, disabled: entry.disabled === true });
  }
  const expanded = expandAll({ values: raw, globals: globalsFor(registry, vault) });
  return meta.map(({ name, disabled }) => {
    const def = registry.variables[name];
    return {
      name,
      value: expanded.get(name) ?? "",
      disabled,
      secret: def?.secret === true,
      groupKey: def?.groupKey,
      example: def?.example,
    };
  });
}

async function classify(
  root: string,
  path: string,
  content: string,
  preview: GeneratePreview,
): Promise<void> {
  const file = Bun.file(join(root, path));
  if (await file.exists()) {
    const existing = await file.text();
    if (existing === content) {
      preview.unchanged.push(path);
      return;
    }
    if (!hasOwnershipMarker(existing)) {
      preview.refused.push(path); // the user took ownership — never overwrite
      return;
    }
  }
  preview.writes.push({ path, content });
}

// Computes every file a generate would write. Pure-ish: reads vaults via the
// passed sessions and the disk only to classify (unchanged/refused). Throws
// VALIDATION (unresolved ref / cycle) before ANY write is possible.
export async function previewGenerate(
  root: string,
  registry: Registry,
  opts: GenerateOpts,
  sessions: ReadonlyMap<string, VaultSession>,
): Promise<GeneratePreview> {
  const preview: GeneratePreview = { writes: [], unchanged: [], refused: [], warnings: [] };
  const exampleDone = new Set<string>();
  for (const target of envTargets(registry.consumers, registry.defaults, opts)) {
    const session = sessions.get(target.vault);
    if (session === undefined) continue; // CLI opens all vaultsNeeded; defensive
    const entries = await scopeEntries(registry, target.consumer, target.vault, session, preview.warnings);
    const def = registry.consumers[target.consumer];
    if (def === undefined) continue;
    const header = disclaimerHeader({ vault: target.vault, consumer: target.consumer });
    const { main, local } = splitSecrets(entries, target.secretsSplit);
    await classify(root, target.relPath, renderEnvContent(main, registry.groups, header), preview);
    if (target.secretsSplit) {
      await classify(root, `${target.relPath}.local`, renderEnvContent(local, registry.groups, header), preview);
    }
    if (def.strategyConfig.example === true && !exampleDone.has(target.consumer)) {
      exampleDone.add(target.consumer);
      // The example documents the full wired surface across vaults: union of
      // names wired anywhere, values-free.
      const names = new Set<string>();
      for (const [name, v] of Object.entries(registry.variables)) {
        for (const byConsumer of Object.values(v.vaultMapping)) {
          if (byConsumer[target.consumer] !== undefined) names.add(name);
        }
      }
      const exampleEntries: RenderEntry[] = [...names].map((name) => {
        const v = registry.variables[name];
        return {
          name,
          value: "",
          disabled: false,
          secret: v?.secret === true,
          groupKey: v?.groupKey,
          example: v?.example,
        };
      });
      const examplePath = join(def.strategyConfig.baseDir, ".env.example");
      const exampleHeader = disclaimerHeader({ consumer: target.consumer });
      await classify(root, examplePath, renderExampleContent(exampleEntries, registry.groups, exampleHeader), preview);
    }
  }
  return preview;
}

export async function applyPreview(root: string, preview: GeneratePreview): Promise<void> {
  for (const w of preview.writes) await writeFileAtomic(root, w.path, w.content);
}

File: go/internal/generate/generate.go
package generate

import (
	"os"
	"path/filepath"
	"sort"

	"github.com/nikrabaev/menv/internal/core"
	"github.com/nikrabaev/menv/internal/registry"
)

// GenerateOpts selects which consumers/vaults to generate for.
type GenerateOpts struct {
	Vault    string // empty = registry default
	Consumer string // empty = all consumers
}

// FileWrite is one file to be written.
type FileWrite struct {
	Path    string
	Content string
}

// GeneratePreview is the result of a preview pass.
type GeneratePreview struct {
	Writes   []FileWrite
	Unchanged []string
	Refused  []string         // files without the ownership marker that would be overwritten
	Warnings []core.PlanIssue
}

// GlobalsFor builds the globals map for a given vault.
func GlobalsFor(reg registry.Registry, vaultName string) map[string]core.GlobalResolution {
	out := map[string]core.GlobalResolution{}
	for name, def := range reg.Globals {
		v, ok := def.Values[vaultName]
		if !ok {
			continue
		}
		if v.Source == "static" {
			out[name] = core.GlobalResolution{Kind: "static", Value: v.Value}
		} else {
			out[name] = core.GlobalResolution{Kind: "runtime"}
		}
	}
	return out
}

// VaultsNeeded returns the distinct vault names required for a generate pass.
func VaultsNeeded(reg registry.Registry, opts GenerateOpts) []string {
	seen := map[string]bool{}
	for _, t := range EnvTargets(reg.Consumers, reg.Defaults, opts) {
		seen[t.Vault] = true
	}
	vaults := make([]string, 0, len(seen))
	for v := range seen {
		vaults = append(vaults, v)
	}
	sort.Strings(vaults)
	return vaults
}

// ScopeEntries fetches and expands all variable values for one (consumer, vault).
func ScopeEntries(
	reg registry.Registry,
	consumer, vaultName string,
	sess core.VaultSession,
	warnings *[]core.PlanIssue,
) ([]RenderEntry, error) {
	rawVals := map[string]string{}
	type meta struct {
		name     string
		disabled bool
	}
	var metas []meta
	for name, def := range reg.Variables {
		entry, ok := def.VaultMapping[vaultName][consumer]
		if !ok {
			continue
		}
		val, _, err := sess.Get(entry.Key)
		if err != nil {
			return nil, err
		}
		if val == "" {
			*warnings = append(*warnings, core.PlanIssue{
				Code:    "MISSING_VALUE",
				Message: `"` + name + `" has no value in vault "` + vaultName + `" (consumer "` + consumer + `") — rendered empty`,
			})
		}
		rawVals[name] = val
		metas = append(metas, meta{name: name, disabled: entry.Disabled})
	}
	globals := GlobalsFor(reg, vaultName)
	expanded, err := core.ExpandAll(core.ExpandInput{Values: rawVals, Globals: globals})
	if err != nil {
		return nil, err
	}
	entries := make([]RenderEntry, len(metas))
	for i, m := range metas {
		def := reg.Variables[m.name]
		entries[i] = RenderEntry{
			Name:     m.name,
			Value:    expanded[m.name],
			Disabled: m.disabled,
			Secret:   def.Secret,
			GroupKey: def.GroupKey,
			Example:  def.Example,
		}
	}
	return entries, nil
}

// readFile is a helper that reads a file; returns os.ErrNotExist if absent.
func readFile(path string) ([]byte, error) {
	return os.ReadFile(path)
}

func classifyWrite(root, relPath, content string, preview *GeneratePreview) error {
	abs := filepath.Join(root, relPath)
	existing, err := os.ReadFile(abs)
	if os.IsNotExist(err) {
		preview.Writes = append(preview.Writes, FileWrite{Path: relPath, Content: content})
		return nil
	}
	if err != nil {
		return err
	}
	if string(existing) == content {
		preview.Unchanged = append(preview.Unchanged, relPath)
		return nil
	}
	if !HasOwnershipMarker(string(existing)) {
		preview.Refused = append(preview.Refused, relPath)
		return nil
	}
	preview.Writes = append(preview.Writes, FileWrite{Path: relPath, Content: content})
	return nil
}

// PreviewGenerate computes every file a generate would write without actually
// writing anything. Reads vault sessions for values; reads disk to classify
// unchanged/refused.
func PreviewGenerate(
	root string,
	reg registry.Registry,
	opts GenerateOpts,
	sessions map[string]core.VaultSession,
) (GeneratePreview, error) {
	preview := GeneratePreview{}
	exampleDone := map[string]bool{}

	for _, target := range EnvTargets(reg.Consumers, reg.Defaults, opts) {
		sess, ok := sessions[target.Vault]
		if !ok {
			continue // CLI opens all VaultsNeeded; defensive skip
		}
		entries, err := ScopeEntries(reg, target.Consumer, target.Vault, sess, &preview.Warnings)
		if err != nil {
			return GeneratePreview{}, err
		}
		def := reg.Consumers[target.Consumer]
		header := DisclaimerHeader(HeaderMeta{Vault: target.Vault, Consumer: target.Consumer})
		main, local := SplitSecrets(entries, target.SecretsSplit)

		if err := classifyWrite(root, target.RelPath, RenderEnvContent(main, reg.Groups, header), &preview); err != nil {
			return GeneratePreview{}, err
		}
		if target.SecretsSplit {
			if err := classifyWrite(root, target.RelPath+".local", RenderEnvContent(local, reg.Groups, header), &preview); err != nil {
				return GeneratePreview{}, err
			}
		}
		if def.StrategyConfig.Example && !exampleDone[target.Consumer] {
			exampleDone[target.Consumer] = true
			// Union of all variable names wired to this consumer across any vault.
			names := map[string]bool{}
			for name, v := range reg.Variables {
				for _, byConsumer := range v.VaultMapping {
					if _, ok := byConsumer[target.Consumer]; ok {
						names[name] = true
					}
				}
			}
			exampleEntries := make([]RenderEntry, 0, len(names))
			for name := range names {
				v := reg.Variables[name]
				exampleEntries = append(exampleEntries, RenderEntry{
					Name:     name,
					Disabled: false,
					Secret:   v.Secret,
					GroupKey: v.GroupKey,
					Example:  v.Example,
				})
			}
			examplePath := filepath.Join(def.StrategyConfig.BaseDir, ".env.example")
			exampleHeader := DisclaimerHeader(HeaderMeta{Consumer: target.Consumer})
			content := RenderExampleContent(exampleEntries, reg.Groups, exampleHeader)
			if err := classifyWrite(root, examplePath, content, &preview); err != nil {
				return GeneratePreview{}, err
			}
		}
	}
	return preview, nil
}
